websockify/token_plugins.py

```diff
@@ -176,6 +176,7 @@ def __init__(self, src):
             # import those ahead of time so we provide error earlier
             import redis
             self._server, self._port = src.split(":")
+            self._client = redis.Redis(host=self._server, port=self._port)
```

Comment:
I'm not sure whether this is safe to do. The server will fork for each request, so multiple requests might be using the same redis connection at the same time. That is likely not supported.

Have you examined this and checked if it is safe? A mere test run is not likely to expose this issue.

Reply:
Yes, this should be safe and more efficient sine we make use of the default redis connection pool, in other to reduce the overhead of creating/destroying connections constantly.

From redis-py documenation about [Connection Pools](https://github.com/redis/redis-py#connection-pools):
```
Behind the scenes, redis-py uses a connection pool to manage connections to a Redis server. By default, each Redis instance you create will in turn create its own connection pool.
```

Before, the code created a new redis client instance for each request with its associated connection pool, but then it did not make any use of the connection pool associated to the instance, because when a new connection came instead of reusing it will create a new instance.

For that reason, just moving the creation of the redis client instance to the contructor will make the trick (similarly to how it would be done for other databases).

Reply:
I have been checking the redis-cli code.

When the Redis client is initialized without arguments, as we do, it creates a connection pool for that client instance. It is done here:
https://github.com/redis/redis-py/blob/3a121bef7bbc5bb5f07b119b0eef2f7527a38eda/redis/client.py#L1033

Then each time a command is executed it uses the current connection (if there is one active) or it gets one from the connection pool:
```
conn = self.connection or pool.get_connection(command_name, **options)
```
https://github.com/redis/redis-py/blob/3a121bef7bbc5bb5f07b119b0eef2f7527a38eda/redis/client.py#L1251

The documentation mentions that in case of using single_connection_client then no connection pool is used and it is not thread safe:
```
single_connection_client:
            if `True`, connection pool is not used. In that case `Redis`
            instance use is not thread safe.
```

If a new process is forked then it will have to instantiate its own TokenRedis and it will have its own connection pool, so you are right at the end in each forked process there will be a new connection pool. The advantage is that if this process does several gets then they will make use of the pool.

I have been running this version during the last months without issues, but we can do further testing. In that case I would propose to remove this part form this pull request and create a dedicated pull request just for this part.

Actually, I think the performance improvement will not be noticeable unless you have a very high number of requests because redis is pretty fast.

Reply:
Removed the connection pool part from current PR and I had to fix also the tests.

Since the connection pool part of the code was based on the changes in PR I will create a new PR when we have this part merged. What do you think?

Reply:
Sounds good.